### DISCUSSION TITLE
Transaction in context manager appears not to commit in specific scenario

### DISCUSSION THREAD

**Discussion by Travis-Brundige** [sqlalchemy/sqlalchemy#11989] [Usage Questions]
In a very specific scenario, it appears that a transaction in a context manager does not commit without explicitly calling the commit method.  I am using:

- An AWS EC2 running Amazon Linux 2
- Python 3.7.16
- SQLAlchemy 1.4.50
- MYSQL 8.0.36

I have both a development and QA environment as described above.  Both environments have the same python script running on them and are referencing identical tables for this post.  I have a python function that runs an update statement shown below.  I pass an engine connection to the function and a table name.  In the function, I define the raw sql update to be executed.  This function and many others like it has been working across many environments and tables.  This is the first time I have seen any odd behavior, in this case apparently not committing (this is my best guess as to what is occurring, not a statement of fact).  

The following conditions provoke the odd behavior:

- There is exactly one record in fn_table
- That record has a process_step = 'pending'
- The python script is invoked from the command line
- The python script is running on my QA environment

Under the above conditions, the record does not have its process_step updated to 'new'.  Variations in conditions where the record(s) do have their process_step(s) updated to 'new':

- If there is more than one record in fn_table, with at least one record with process_step = 'pending', in this case all pending records update to new
- If I run the python script using vscode: run - start debugging or run - run without debugging (even if there is only one record in the table), this works whether or not any breakpoints are set
- If I run in my development environment, on an identical table to the QA environment, even if the table has only one record, it updates correctly

I have placed log statements inside the context manager, and have confirmed that the query being executed is the expected query, and that query does update a single record if manually executed in QA.  

The original function:

```
def update_pending_to_new(fn_connection, fn_table):
    """Updates existing pending records to new, when no other record is present."""
    update = f"""UPDATE {fn_table} f
    INNER JOIN (SELECT batch_id
    FROM {fn_table} f1
    INNER JOIN (
    SELECT program_id
    FROM {fn_table} f2
    GROUP BY program_id
    HAVING COUNT(*) = 1
    ) p
    ON f1.program_id = p.program_id
    WHERE f1.process_step = 'pending'
    ) u
    ON f.batch_id = u.batch_id
    SET f.process_step = 'new'
    WHERE u.batch_id IS NOT NULL
    """
    try:
        with fn_connection.begin():
            fn_connection.execute(text(update))
    except RuntimeError:
        log(
            f"Error updating existing pending records to new in table {fn_table}: {sys.exc_info()[0]}"
        )
        sys.exit(177)
    else:
        log(
            f"Successfully updated existing pending records to new in table {fn_table}."
        )

```

The updated function that consistently works as expected and does not exhibit any odd behavior (under the specified conditions, the record is updated to have a process_step = 'new').  The only changes I have made are to name the context manager and to explicitly commit using the transaction commit method.  

```
def update_pending_to_new(fn_connection, fn_table):
    """Updates existing pending records to new, when no other record is present."""
    update = f"""UPDATE {fn_table} f
    INNER JOIN (SELECT batch_id
    FROM {fn_table} f1
    INNER JOIN (
    SELECT program_id
    FROM {fn_table} f2
    GROUP BY program_id
    HAVING COUNT(*) = 1
    ) p
    ON f1.program_id = p.program_id
    WHERE f1.process_step = 'pending'
    ) u
    ON f.batch_id = u.batch_id
    SET f.process_step = 'new'
    WHERE u.batch_id IS NOT NULL
    """
    try:
        with fn_connection.begin() as transaction:
            fn_connection.execute(text(update))
            transaction.commit()
    except RuntimeError:
        log(
            f"Error updating existing pending records to new in table {fn_table}: {sys.exc_info()[0]}"
        )
        sys.exit(177)
    else:
        log(
            f"Successfully updated existing pending records to new in table {fn_table}."
        )

```

There are no error messages, and I do get the successful log message in all cases.  For reference, the documentation states, 

'The block managed by each .begin() method has the behavior such that the transaction is committed when the block completes. If an exception is raised, the transaction is instead rolled back, and the exception propagated outwards.'

from https://docs.sqlalchemy.org/en/14/core/connections.html and it appears to me that in this case the described behavior is not occurring.  

This is my first post seeking help in a GitHub discussion.  I would be happy to restate anything that is unclear or to otherwise amend my post.  

Thank you for your time.

**Comment by CaselIT** [comment]
Hi,

This seems quite odd, and I would assume the issue is in the data. I would start by:
- enabling echo in the engine so you can see what gets executed
- checking the value of rowcount returned by the execute function
